File: src/web/middleware/pydantic_models/sub_task.py
import json
import time
from pydantic import BaseModel
from typing import Any
from pydantic_models.action import Action
from pydantic_models.submodule_address import SubmoduleAddress


class SubTask(BaseModel):
    id: str
    name: str
    status: str
    assignee_name: str
    parent_id: str
    requires_task_id: str | None
    is_part_of_monolith: bool
    target_id: str
    priority: int
    earliest_start_time: int
    requirements: dict[str, Any]
    action: Action

    @classmethod
    def from_json(cls, json_data: dict[str, Any]) -> "SubTask":
        return cls(
            id=json_data["id"],
            name=json_data["name"],
            status=json_data["status"],
            assignee_name=json_data["assignee_name"],
            parent_id=json_data["parent_id"],
            requires_task_id=json_data["requires_task_id"],
            is_part_of_monolith=json_data["is_part_of_monolith"],
            target_id=json_data["target_id"],
            priority=json_data["priority"],
            earliest_start_time=json_data["earliest_start_time"],
            requirements=json_data["requirements"],
            action=Action.from_json(json_data["action"]),
        )

    @classmethod
    def from_database_row(cls, row: list[Any]) -> "SubTask":
        return cls(
            id=row[0],
            name=row[1],
            status=row[2],
            assignee_name=row[3],
            parent_id=row[4],
            requires_task_id=row[5],
            is_part_of_monolith=row[6],
            target_id=row[7],
            priority=row[8],
            earliest_start_time=row[9],
            requirements=json.loads(row[10]),
            action=Action.from_json(json.loads(row[11])),
        )

    def contains_submodule_process_action(self) -> bool:
        action = self.action
        while action:
            if action.name == "submodule_process":
                return True
            action = action.subaction
        return False

    def write_submodule_address(self, submodule_address: SubmoduleAddress) -> None:
        submodule_address_json = submodule_address.to_json()
        self.requirements["submodule_address"] = submodule_address_json
        action = self.action
        while action:
            if action.name == "submodule_process":
                action.parameters["submodule_address"] = submodule_address_json
            action = action.subaction

    def to_json(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "name": self.name,
            "status": self.status,
            "assignee_name": self.assignee_name,
            "parent_id": self.parent_id,
            "requires_task_id": self.requires_task_id,
            "is_part_of_monolith": self.is_part_of_monolith,
            "target_id": self.target_id,
            "priority": self.priority,
            "earliest_start_time": self.earliest_start_time,
            "requirements": self.requirements,
            "action": self.action.to_json(),
        }

    def to_robot_task_request(self) -> dict[str, Any]:
        return {
            "type": "robot_task_request",
            "robot": self.assignee_name,
            "fleet": "deliveryRobot",
            "request": {
                "unix_millis_earliest_start_time": int(time.time() * 1000),
                "unix_millis_request_time": int(time.time() * 1000),
                "priority": {"value": 0, "type": "binary"},
                "category": "compose",
                "description": {
                    "category": "custom_action",
                    "phases": [self.__create_task_phase()],
                },
                "requester": "task_system",
            },
        }

    def __create_task_phase(self) -> dict[str, Any]:
        return {"activity": self.__create_task_sequence()}

    def __create_task_sequence(self) -> dict[str, Any]:
        return {
            "category": "sequence",
            "description": {
                "activities": [self.__create_activity()],
            },
        }

    def __create_activity(self) -> dict[str, Any]:
        return {
            "category": "pickup",
            "description": {
                "place": self.target_id,
                "handler": "submodule_dispenser",
                "payload": {
                    "sku": json.dumps(self.action.to_json()),
                    "quantity": 1,
                },
            },
        }

File: src/web/frontend/robot_apps_flutter/packages/middleware_api_utilities/lib/src/models/subtask.dart
import 'package:middleware_api_utilities/middleware_api_utilities.dart';
import 'package:middleware_api_utilities/src/models/robot_action.dart';
import 'package:middleware_api_utilities/src/models/submodule_address.dart';
import 'package:uuid/uuid.dart';

class SubTask {
  SubTask({
    required this.id,
    required this.name,
    required this.status,
    required this.assigneeName,
    required this.parentID,
    required this.requiresTaskID,
    required this.isPartOfMonolith,
    required this.targetID,
    required this.priority,
    required this.earliestStartTime,
    required this.requirements,
    required this.action,
  });

  factory SubTask.fromJson(Map<String, dynamic> json) {
    return SubTask(
      id: json['id'] as String,
      name: json['name'] as String,
      status: json['status'] as String,
      assigneeName: json['assignee_name'] as String,
      parentID: json['parent_id'] as String,
      requiresTaskID: json['requires_task_id'] as String?,
      isPartOfMonolith: json['is_part_of_monolith'] as bool,
      targetID: json['target_id'] as String,
      priority: json['priority'] as int,
      earliestStartTime: json['earliest_start_time'] as int,
      requirements: json['requirements'] as Map<String, dynamic>,
      action: RobotAction.fromJson(json['action'] as Map<String, dynamic>),
    );
  }

  factory SubTask.submoduleProcess({
    required String id,
    required String name,
    required String parentID,
    required String? requiresTaskID,
    required String targetID,
    required List<String> requiredUserIDs,
    required List<String> requiredUserGroups,
    required Map<String, int> itemsByChange,
  }) {
    return SubTask(
      id: id,
      name: name,
      status: 'pending',
      assigneeName: '',
      requiresTaskID: requiresTaskID,
      isPartOfMonolith: false,
      parentID: parentID,
      targetID: targetID,
      priority: 0,
      earliestStartTime: DateTime.now().millisecondsSinceEpoch ~/ 1000,
      requirements: {
        'required_user_ids': requiredUserIDs,
        'required_user_groups': requiredUserGroups,
      },
      action: RobotAction(
        id: const Uuid().v4(),
        name: 'submodule_process',
        status: 'pending',
        parameters: {
          'items_by_change': itemsByChange,
        },
        subaction: null,
      ),
    );
  }

  factory SubTask.assignedSubmoduleProcess({
    required String id,
    required String parentID,
    required String? requiresTaskID,
    required String targetID,
    required List<String> requiredUserIDs,
    required List<String> requiredUserGroups,
    required Map<String, int> itemsByChange,
    required SubmoduleAddress submoduleAddress,
    required String assigneeName,
  }) {
    final completeSubmoduleAddress = {'submodule_address': submoduleAddress.toJson()};
    completeSubmoduleAddress['submodule_address']!['robot_name'] = assigneeName;
    return SubTask(
      id: id,
      name: 'Zustellung',
      status: 'pending',
      assigneeName: assigneeName,
      requiresTaskID: requiresTaskID,
      isPartOfMonolith: false,
      parentID: parentID,
      targetID: targetID,
      priority: 0,
      earliestStartTime: DateTime.now().millisecondsSinceEpoch ~/ 1000,
      requirements: {
        'required_user_ids': requiredUserIDs,
        'required_user_groups': requiredUserGroups,
        ...completeSubmoduleAddress,
      },
      action: RobotAction(
        id: const Uuid().v4(),
        name: 'submodule_process',
        status: 'pending',
        parameters: {
          'items_by_change': itemsByChange,
          ...completeSubmoduleAddress,
        },
        subaction: null,
      ),
    );
  }

  final String id;
  final String name;
  final String status;
  final String assigneeName;
  final String parentID;
  final String? requiresTaskID;
  final bool isPartOfMonolith;
  final String targetID;
  final int priority;
  final int earliestStartTime;
  final Map<String, dynamic> requirements;
  final RobotAction action;

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'status': status,
      'assignee_name': assigneeName,
      'parent_id': parentID,
      'requires_task_id': requiresTaskID,
      'is_part_of_monolith': isPartOfMonolith,
      'target_id': targetID,
      'priority': priority,
      'earliest_start_time': earliestStartTime,
      'requirements': requirements,
      'action': action.toJson(),
    };
  }
}
